Cannot find module 'yup' or its corresponding type declarations

**Versions**

- vee-validate": "^4.4.0-alpha.2 and 4.3.6
- vue: 3.0.5

**Describe the bug**

I'm using vite/vue 3 and when I'm building my project, I got this error:
```
node_modules/vee-validate/dist/vee-validate.d.ts:3:66 - error TS2307: Cannot find module 'yup' or its corresponding type declarations.

3 import { SchemaOf, AnySchema, AnyObjectSchema, BaseSchema } from 'yup';
```

I think that `yup` should be added as `peerDependencies` in  `node_modules/vee-validate/package.json`  just like `vue` package
In order to fix my issue I installed it as dev dependencies in my package.

Notice: I'm not using yup at all in my validatation form process.

Sorry @logaretm , I should have done it since the beginning.

```
git clone git@github.com:Lord-Y/vite-vee-validate-debug.git
npm i
npm run build
```

So with the initial setup with vee-validate in `main.ts` (no component created to use vee-validate) the build is failing like so:

```
> vue-tsc --noEmit && vite build
node_modules/vee-validate/dist/vee-validate.d.ts:3:66 - error TS2307: Cannot find module 'yup' or its corresponding type declarations.

3 import { SchemaOf, AnySchema, AnyObjectSchema, BaseSchema } from 'yup';
                                                                   ~~~~~
Found 1 error.
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! vite-vee-validate-debug@0.0.0 build: `vue-tsc --noEmit && vite build`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the vite-vee-validate-debug@0.0.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```
Doc: https://vee-validate.logaretm.com/v4/v4/guide/global-validators#vee-validaterules

Adding yup as dependency (ie`npm install yup@next` or `npm install -D yup@next`) does not solve the problem...
I'm using vue v3.2.25, vite v2.7.2, vee-validate v4.5.8 and yup v1.0.0-beta.2

    npm run build
    > xxxxxxxx@0.0.1 build
    > vue-tsc --noEmit && vite build
    node_modules/vee-validate/dist/vee-validate.d.ts:3:38 - error TS2724: '"yup"' has no exported member named 'SchemaOf'. Did you mean 'Schema'?
    3 import { AnySchema, AnyObjectSchema, SchemaOf, BaseSchema } from 'yup';
                                           ~~~~~~~~
    node_modules/vee-validate/dist/vee-validate.d.ts:3:48 - error TS2724: '"yup"' has no exported member named 'BaseSchema'. Did you mean 'DateSchema'?
    3 import { AnySchema, AnyObjectSchema, SchemaOf, BaseSchema } from 'yup';
                                                     ~~~~~~~~~~
    Found 2 errors.

So the propse workaround does not work

--- a/packages/vee-validate/src/types.ts
+++ b/packages/vee-validate/src/types.ts
@@ -1,5 +1,4 @@
 import { ComputedRef, Ref } from 'vue';
-import { SchemaOf, AnySchema, AnyObjectSchema } from 'yup';
 import { FieldValidationMetaInfo } from '../../shared';
 
 export type GenericFormValues = Record<string, unknown>;
@@ -9,7 +8,14 @@ export interface ValidationResult {
   valid: boolean;
 }
 
-export type YupValidator = AnySchema | AnyObjectSchema;
+export type YupValidator<TValue = any> = { validate(value: TValue, options: Record<string, any>): Promise<TValue> };
+
+export interface YupValidationError extends Error {
+  value: any;
+  path?: string;
+  errors: string[];
+  inner: YupValidationError[];
+}
 
 export type Locator = { __locatorRef: string } & ((values: GenericFormValues) => unknown);
 
@@ -178,7 +184,7 @@ export interface PrivateFormContext<TValues extends Record<string, any> = Record
   fieldsByPath: Ref<FieldPathLookup>;
   fieldArrays: PrivateFieldArrayContext[];
   submitCount: Ref<number>;
-  schema?: MaybeRef<RawFormSchema<TValues> | SchemaOf<TValues> | undefined>;
+  schema?: MaybeRef<RawFormSchema<TValues> | YupValidator<TValues> | undefined>;
   errorBag: Ref<FormErrorBag<TValues>>;
   errors: ComputedRef<FormErrors<TValues>>;
   meta: ComputedRef<FormMeta<TValues>>;

--- a/packages/vee-validate/src/useField.ts
+++ b/packages/vee-validate/src/useField.ts
@@ -10,7 +10,6 @@ import {
   nextTick,
   getCurrentInstance,
 } from 'vue';
-import { BaseSchema } from 'yup';
 import { klona as deepCopy } from 'klona/full';
 import isEqual from 'fast-deep-equal/es6';
 import { validate as validateValue } from './validate';
@@ -59,8 +58,7 @@ export type RuleExpression<TValue> =
   | Record<string, unknown>
   | GenericValidateFunction
   | GenericValidateFunction[]
-  | YupValidator
-  | BaseSchema<TValue>
+  | YupValidator<TValue>
   | undefined;
 
 /**

--- a/packages/vee-validate/src/useForm.ts
+++ b/packages/vee-validate/src/useForm.ts
@@ -14,7 +14,6 @@ import {
   watchEffect,
 } from 'vue';
 import isEqual from 'fast-deep-equal/es6';
-import type { SchemaOf } from 'yup';
 import { klona as deepCopy } from 'klona/full';
 import {
   FieldMeta,
@@ -55,7 +54,7 @@ import { _useFieldValue } from './useFieldState';
 
 interface FormOptions<TValues extends Record<string, any>> {
   validationSchema?: MaybeRef<
-    Record<keyof TValues, GenericValidateFunction | string | Record<string, any>> | SchemaOf<TValues> | undefined
+    Record<keyof TValues, GenericValidateFunction | string | Record<string, any>> | any | undefined
   >;
   initialValues?: MaybeRef<TValues>;
   initialErrors?: Record<keyof TValues, string | undefined>;

--- a/packages/vee-validate/src/validate.ts
+++ b/packages/vee-validate/src/validate.ts
@@ -1,8 +1,14 @@
-import type { SchemaOf, ValidationError } from 'yup';
 import { resolveRule } from './defineRule';
 import { isLocator, normalizeRules, isYupValidator, keysOf, getFromPath } from './utils';
 import { getConfig } from './config';
-import { ValidationResult, GenericValidateFunction, YupValidator, FormValidationResult, RawFormSchema } from './types';
+import {
+  ValidationResult,
+  GenericValidateFunction,
+  YupValidator,
+  FormValidationResult,
+  RawFormSchema,
+  YupValidationError,
+} from './types';
 import { isCallable, FieldValidationMetaInfo } from '../../shared';
 
 /**
@@ -137,7 +143,7 @@ async function validateFieldWithYup(value: unknown, validator: YupValidator, opt
       abortEarly: opts.bails ?? true,
     })
     .then(() => [])
-    .catch((err: ValidationError) => {
+    .catch((err: YupValidationError) => {
       // Yup errors have a name prop one them.
       // https://github.com/jquense/yup#validationerrorerrors-string--arraystring-value-any-path-string
       if (err.name === 'ValidationError') {
@@ -223,13 +229,13 @@ function fillTargetValues(params: unknown[] | Record<string, unknown>, crossTabl
 }
 
 export async function validateYupSchema<TValues>(
-  schema: SchemaOf<TValues>,
+  schema: YupValidator<TValues>,
   values: TValues
 ): Promise<FormValidationResult<TValues>> {
-  const errorObjects: ValidationError[] = await (schema as YupValidator)
+  const errorObjects: YupValidationError[] = await (schema as YupValidator)
     .validate(values, { abortEarly: false })
     .then(() => [])
-    .catch((err: ValidationError) => {
+    .catch((err: YupValidationError) => {
       // Yup errors have a name prop one them.
       // https://github.com/jquense/yup#validationerrorerrors-string--arraystring-value-any-path-string
       if (err.name !== 'ValidationError') {

--- a/packages/zod/src/index.ts
+++ b/packages/zod/src/index.ts
@@ -1,5 +1,4 @@
 import type { ZodObject, ZodType, ZodTypeDef, TypeOf as ZodTypeOf, ZodRawShape, ZodEffects } from 'zod';
-import type { BaseSchema, SchemaOf } from 'yup';
 import { isIndex } from '../../shared';
 
 /**
@@ -7,7 +6,7 @@ import { isIndex } from '../../shared';
  */
 export function toFieldValidator<TValue = unknown, TDef extends ZodTypeDef = ZodTypeDef, TInput = TValue>(
   zodSchema: ZodType<TValue, TDef, TInput> | ZodEffects<ZodType<TValue, TDef, TInput>>
-): BaseSchema<TValue> {
+): any {
   return {
     async validate(value: TValue) {
       const result = await zodSchema.safeParseAsync(value);
@@ -21,7 +20,7 @@ export function toFieldValidator<TValue = unknown, TDef extends ZodTypeDef = Zod
 
       throw error;
     },
-  } as BaseSchema<TValue>;
+  } as any;
 }
 
 interface AggregatedZodError {
@@ -39,7 +38,7 @@ type ToBaseTypes<TShape extends ZodRawShape> = {
 export function toFormValidator<
   TShape extends ZodRawShape,
   TValues extends Record<string, unknown> = ToBaseTypes<TShape>
->(zodSchema: ZodObject<TShape> | ZodEffects<ZodObject<TShape>>): SchemaOf<TValues> {
+>(zodSchema: ZodObject<TShape> | ZodEffects<ZodObject<TShape>>) {
   return {
     async validate(value: TValues) {
       const result = await zodSchema.safeParseAsync(value);
@@ -57,7 +56,7 @@ export function toFormValidator<
 
       throw error;
     },
-  } as SchemaOf<TValues>;
+  } as any;
 }
 
 /**

